src/java/org/apache/cassandra/net/InboundSockets.java

```diff
@@ -99,15 +99,28 @@ private InboundSocket(InboundConnectionSettings settings)
                     throw new IllegalStateException();
                 binding = InboundConnectionInitiator.bind(settings, connections, pipelineInjector);
             }
-
-            return binding.addListener(ignore -> {
+            // isOpen is defined as "listen.isOpen", but this is set AFTER the binding future is set
+            // to make sure the future returned does not complete until listen is set, need a new
+            // future to replicate "Future.map" behavior.
+            AsyncChannelPromise promise = new AsyncChannelPromise(binding.channel());
+            binding.addListener(f -> {
+                if (!f.isSuccess())
+                {
+                    synchronized (this)
+                    {
+                        binding = null;
+                    }
+                    promise.setFailure(f.cause());
+                    return;
+                }
                 synchronized (this)
                 {
-                    if (binding.isSuccess())
-                        listen = binding.channel();
+                    listen = binding.channel();
                     binding = null;
                 }
+                promise.setSuccess(null);
             });
+            return promise;
```

Comment:
So to summarize my understanding, waiting for `binding`, the previously returned `Future` was insufficient, because the body of the listener is not guaranteed to run before it presents itself as complete. So (taking care to protect the shared mutable state of `binding` and `listen`), we create a separate `promise` whose completion guarantees that the state we need to change has already been changed. If so, +1 :)